### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by sgaamuwa**

### CODE REVIEW

**Comment by njirap** [comment]
- Squash squashable commits eg.. from `ad942d1` to `44b8627`
- Improve on commit titles and bodies
- Don't push commented code
- Don't be redundant in your imports e.g. `test_amity.py` `import unittest` and `from unittest.mock import patch, PropertyMock, MagicMock, Mock`
- Install and work with pep8
- In `test_amity`, for testing of the save_state to the db you can mock the db call and assert that it has been called with the parameters
- Checkout python's `tempfile` library for creating temporary files and directories instead of having permanent ones in the test
- In `app.py` you have imported `os` but you are not using it
- `app.py` `do_reallocate_person` has nested loops. They can be made more efficient by maybe using dictionaries or comprehensions
- `amity.py` Lines 26 & 27...maybe no need for variable assignment `room`
- `amity.py` Line 65... Maybe `if staff_id in Amity.fellows.keys() + Amity.staff.keys()` . Repeat for all other ifs in the previous format
- Amity.assign_room should probably be Amity.assign_office...i.e make variables make sense
- amity.py `add_person` if statement can be refactored
  L52...we probably don't need a default return message
  L79...`if office is not "" or None` ...Maybe work on all if statements
  Maybe `reallocate` and `reallocate_ls` can be squashed and the new function takes a third parameter of room type
  `print_allocations` and `print_unallocated` maybe develop the entire text first then decide whether to output to terminal or write to file

**Comment by sgaamuwa** [comment]
Changes made except for:
- mocking calls to the database in save state
- using a tempfile in test_amity, it was not possible to specify the directory for the file so it kept conflicting with the code, opted to leave it as it was 
- `amity.py` line 65, dict.keys() lists can't be summed up like normal lists, so that raised an error

File: classes/amity.py [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+        #combine the two reallocate functions 
+        """assigns a person to the specified room if it is free"""
+        if room_type == "office":
+            room_dict = Amity.offices
+            current_room = person.allocated_office
+        elif room_type ==  "livingspace":
+            room_dict = Amity.livingspaces
+            current_room = person.allocated_livingspace
+        #check for the room in the rooms list 
+        if room_name in room_dict.keys():
```
Comment (duboviy) [inline classes/amity.py:154]: Same here

File: app.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+    @docopt_cmd
+    def do_print_room(self, args):
+        """
+        Prints out a single specified room and the people allocated to iter
+        
+        Usage: print_room <room_name>
+        """
+        room = args["<room_name>"]
+
+        if room in Amity.offices.keys():
```
Comment (duboviy) [inline app.py:183]: Same here

File: app.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+            Amity.offices[room].print_room()
+        elif room in Amity.livingspaces.keys():
```
Comment (duboviy) [inline app.py:185]: Same here

File: app.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+        Reallocates a person to a new room using their unique identifier
+        When loaded into the system, each person has a unique staff_id number
+        This is used to reallocate them to specific rooms
+
+        Usage: reallocate_person <person_identifier> <new_room_name>
+        """
+        person = args["<person_identifier>"]
+        room = args["<new_room_name>"]
+
+        if person in Amity.fellows.keys():
```
Comment (duboviy) [inline app.py:113]: Same here

File: app.py [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+            perp = Amity.fellows[person]
+        elif person in Amity.staff.keys():
```
Comment (duboviy) [inline app.py:115]: Same here

File: app.py [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+            perp = Amity.staff[person]
+        if room in Amity.offices.keys():
```
Comment (duboviy) [inline app.py:117]: Same here

File: app.py [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+            print(Amity.reallocate(perp, room, "office"))
+        elif room in Amity.livingspaces.keys():
```
Comment (duboviy) [inline app.py:119]: Same here

File: app.py [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+            room_type = "livingspace"
+            if person in Amity.staff.keys():
```
Comment (duboviy) [inline app.py:121]: Same here

File: tests/test_amity.py [hunk 9]
```diff
@@ -1,50 +1,190 @@
...
-        self.assertEqual(LivingSpace.objects.count, 1)
+        #test that create room creates both offices and livingspaces
+        #query the dictionaries to see the rooms are created
+        with mock.patch("classes.amity.Office") as patched_office:
+            Amity.create_room("Oculus", "office")
+            self.assertIn("Oculus", Amity.offices.keys())
+        #do the same for livingspace
+        with mock.patch("classes.amity.LivingSpace") as patched_lspace:
+            Amity.create_room("Python", "livingspace")
+            self.assertIn("Python", Amity.livingspaces.keys())
```
Comment (duboviy) [inline tests/test_amity.py:21]: Same here

File: tests/test_amity.py [hunk 10]
```diff
@@ -1,50 +1,190 @@
...
     def test_create_room(self):
-        Amity.create_room("Oculus", "office")
-        self.assertEqual(Office.objects.count, 1)
-        Amity.create_room("Python", "living space")
-        self.assertEqual(LivingSpace.objects.count, 1)
+        #test that create room creates both offices and livingspaces
+        #query the dictionaries to see the rooms are created
+        with mock.patch("classes.amity.Office") as patched_office:
+            Amity.create_room("Oculus", "office")
+            self.assertIn("Oculus", Amity.offices.keys())
```
Comment (duboviy) [inline tests/test_amity.py:17]: Same here

File: classes/amity.py [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+                                                                person.last_name)
+        else:
+            return "{} {} added and assigned to {}".format(person.first_name, 
+                                                            person.last_name, 
+                                                            person.allocated_office)
+
+    def validate_staff_id(staff_id):
+        """determines that the staff id is not in the system"""
+
+        if staff_id in Amity.fellows.keys() or staff_id in Amity.staff.keys():
```
Comment (duboviy) [inline classes/amity.py:60]: Same here

**Review by duboviy — CHANGES_REQUESTED** [review]
- Please do not use `your_dict.keys()` to iterate over a dictionary:
```python
# Bad
for key in your_dict.keys():
    ...

# Better
for key in your_dict:
    ...
```

File: classes/amity.py (hunk 11 above)
Comment (sgaamuwa) [reply classes/amity.py:60]: Thank you, did not know it iterates over the keys and not the values, much appreciated

Comment (duboviy) [reply classes/amity.py:60]: You are welcome!
I'm glad that my comment was useful for you ;)

**Pull request merged**